Sprint 9 solution http api

shinket7 opened the pull request.

shinket7 committed:
Add gson jar to project.

shinket7 committed:
Create `HttpTaskServer` without handlers.

shinket7 committed:
Create `BaseHttpHandler`.

Create methods: `sendText()`, `sendBadRequest()`, `sendNotFound()`, `sendNotAllowed()`, `sendHasOverlaps()`.

shinket7 committed:
Add task manager creation when server starts.

shinket7 committed:
Add task manager field to `BaseHttpHandler`.

shinket7 committed:
Add `NotFoundException` when issue is not found by id.

shinket7 committed:
Add gson field to `BaseHeepHandler`.

shinket7 committed:
Add tasks GET handlers.

shinket7 committed:
Add task GET and POST handlers.

shinket7 committed:
Add overlaps check logic to tasks api.

shinket7 committed:
Fix prioritized tasks tree set issue.

Fix issue when task with different start time but same task id was considered to be another task instead of the same.

shinket7 committed:
Create task delete handler.

shinket7 committed:
Create epic handler.

shinket7 committed:
Create subtask handler.

shinket7 committed:
Create user handler.

shinket7 committed:
Create task handler tests.

shinket7 committed:
Add field absence checks for POST methods.

shinket7 committed:
Take out inner auxiliary test classes.

shinket7 committed:
Create subtask handler tests.

shinket7 committed:
Fix epics handler POST method.

Fix issue when `taskId` field of `-1` was considered as id for search.

shinket7 committed:
Create epic handler tests.

shinket7 committed:
Create user handler tests.

shinket7 committed:
Code style fix in `BaseHttpHandler`.

shinket7 committed:
Code style fix in `FileBackedTaskManager`.

shinket7 committed:
Code style fix in `InMemoryTaskManager`.

avfyodorov commented on src/tracker/api/DurationTypeAdapter.java:
```diff
@@ -0,0 +1,22 @@
+import java.io.IOException;
+import java.time.Duration;
+
+public class DurationTypeAdapter extends TypeAdapter<Duration> {
```
Самодельные адаптеры принято оформлять как классы и хранить отдельно в пакете, например, adapters

shinket7 replied:
Создал отдельный пакет для них и перенёс туда

avfyodorov commented on src/tracker/api/BaseHttpHandler.java:
```diff
@@ -0,0 +1,60 @@
+import java.time.Duration;
+import java.time.LocalDateTime;
+
+public class BaseHttpHandler {
```
На Ваше усмотрение. 

Можно было бы провести рефакторинг, в базовом классе (BaseHttpHandler) оставить обработчик и методы-заглушки, а в классах наследниках переопределять только эти методы (processGet, processPost….), если они нужны. Такое решение лучше соответствует принципам OOP 
```java
public  class BaseHandler implements HttpHandler {
......................

public void handle(HttpExchange exchange) throws IOException { 
       String method = exchange.getRequestMethod();
           switch (method) {
               case "GET":
                   processGet(exchange);
                   break;
               case "POST":
                   processPost(exchange);
                   break;
               case "DELETE":
                   processDelete(exchange);
                   break;
               default:
                   writeToUser(exchange, "Данный метод не предусмотрен");
           }
           
    protected void processGet(....) {
        sendMethodNotAllowed(...);
   }
   protected void processPost(....) {
        sendMethodNotAllowed(...);
   }
.......................
И здесь же общие методы для классов-наследников. 

```
   Далее в каждом из наследников мы будем переопределять только нужные для этого наследника методы  processGet, processPost и т.д. и реализовывать в них нужную нам логику)

shinket7 replied:
Реализовал разделение по методам внутри хендлера

avfyodorov commented on src/tracker/api/TaskHandler.java:
```diff
@@ -0,0 +1,205 @@
+import java.time.LocalDateTime;
+import java.util.List;
+
+public class TaskHandler extends BaseHttpHandler implements HttpHandler {
```
Для задач приблизительно так:
```java
public class TasksHandler extends BaseHandler {
......................
    public TasksHandler(........................
    
    @Override
    protected void processGet(HttpExchange exchange, String path) throws IOException {
  .........................
    @Override
    protected void processPost(HttpExchange exchange) throws IOException {
 ...............
    @Override
    protected void processDelete(HttpExchange exchange, String path) throws IOException {
.......................
```

avfyodorov commented on src/tracker/api/TaskHandler.java:
```diff
@@ -0,0 +1,205 @@
+    public TaskHandler(TaskManager taskManager) {
+        super(taskManager);
+    }
+
```
А для отсортированного по времени списка достаточно будет переопределить всего один метод: 

```java
public class PrioritizedHandler extends BaseHandler {

    @Override
    protected void processGet(HttpExchange exchange, String path) throws IOException {
..................
```
.---------------------

Аналогично определяются классы-наследники и для эпиков, подзадач, истории.

avfyodorov commented on src/tracker/api/HttpTaskServer.java:
```diff
@@ -0,0 +1,43 @@
+        server.createContext("/tasks", taskHandler);
+        server.createContext("/subtasks", subtaskHandler);
+        server.createContext("/epics", epicHandler);
+        server.createContext("/", userHandler);
```
Не то, чтобы я придирался, но всё же... 
Лучше сделать два разных обработчика для истории и отсортированного списка задач, не надо смешивать логику. 
И уж точно это не юзер. )
`/history`
`/prioritized`

shinket7 replied:
Разделил на два хендлера: для истории и для приоритезированных

avfyodorov reviewed (commented):
Добрый вечер, Виталий! 

Очень хорошая работа, от меня всего пара уточнений.

shinket7 committed:
Move adapters from api package to a new one.

shinket7 committed:
Refactor task handler and base handler.

Move `handle()` to base handler and split its logic to `processGet()`, `processPost()`, `processDelete()`.

shinket7 committed:
Refactor subtask handler and base handler.

Move `handle()` to base handler and split its logic to `processGet()`, `processPost()`, `processDelete()`.

shinket7 committed:
Refactor epic handler and base handler.

Move `handle()` to base handler and split its logic to `processGet()`, `processPost()`, `processDelete()`.

shinket7 committed:
Refactor user handler.

Split user handler to history handler and prioritized handler.

avfyodorov reviewed (commented):
Добрый день, Виталий!

Замечаний нет.
Работа принята.
